### ISSUE TITLE
false error with simple nested structure

### ISSUE THREAD

**Issue by zoranbosnjak** [python/mypy#17143]
**Bug Report**

`mypy` reports error where it is not suppose to.

**To Reproduce**

```python
from typing import *                                                                                       
                                                                                                           
# types                                                                                                    
MyDict = TypedDict('MyDict', {"a" : int, "b" : str}, total=False)                                          
MyUnion = Union[                                                                                           
    Tuple[Literal["arg_int"], int],                                                                        
    Tuple[Literal["arg_str"], str],                                                                        
    Tuple[Literal["arg_dict"], MyDict]                                                                     
]                                                                                                          
                                                                                                           
# variables                                                                                                
var1 : MyUnion = ('arg_int', 1)                                                                            
var2 : MyUnion = ('arg_str', 'test')                                                                       
var3 : MyUnion = ('arg_dict', {'a': 1})    
```

**Expected Behavior**

No error is expected in this case.

**Actual Behavior**

The following error is reported:

```
error: Incompatible types in assignment (expression has type "tuple[str, dict[str, int]]", variable has type "tuple[Literal['arg_int'], int] | tuple[Literal['arg_str'], str] | tuple[Literal['arg_dict'], MyDict]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.9.0
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.11.8

Also tryed with mypy playground with all latest versions and got the same result.

**Comment by lbernwald-tng** [comment]
I have a similar issue. Using the following code
```
from typing import *

# types
MyUnion = (
    Dict[int, str] |
    Dict[int, Dict[str, str | None]]
)

# variables
var1: MyUnion = {404: {"description": "Resource was not found"}}
```
I get the error
```
error: Incompatible types in assignment (expression has type "dict[int, dict[str, str]]", variable has type "dict[int, str] | dict[int, dict[str, str | None]]")  [assignment]
```
My environment

- mypy version: 1.13.0
- python version: 3.11.10

It also doesn't work on the mypy plaground using the latest versions.
